python/pyarrow/io.pxi

```diff
@@ -1479,13 +1464,19 @@ def input_stream(source, compression='detect'):
         raise TypeError("pa.input_stream() called with instance of '{}'"
                         .format(source.__class__))
 
-    if compression_type != CompressionType_UNCOMPRESSED:
-        stream = CompressedInputStream.create(stream, compression_type)
+    if compression == 'detect':
+        compression = _detect_compression(source_path)
+
+    if compression is not None:
+        stream = CompressedInputStream(stream, compression)
+
+    if buffer_size is not None:
+        stream = BufferedInputStream(stream, buffer_size)
```

Comment:
I guess this is the preferred wrapping order: `Buffered[Compressed[Raw]]`

Reply:
A compressed stream already has an internal buffer, so it's probably pointless to wrap it in another buffered reader. So I'd say `Compressed[Raw]`.

(you may want to run a quick benchmark to check that out)

Reply:
I think this would violate the intent of the user's passed buffer_size. One primary purpose of buffering is to mitigate read latency, which can be high for remote file systems. This would suggest that Compressed[Buffered[Raw]] is the correct wrapping order. Benchmarks with simulated read latency can obviously prove which is correct